common/check_internal.h

```diff
@@ -5,73 +5,128 @@
+[[noreturn, gnu::cold]] auto CheckFail(const char* kind, const char* file,
+                                       int line, const char* condition_str,
+                                       const char* format_str, Ts&&... values)
+    -> void {
+  if constexpr (sizeof...(values) == 0) {
+    CheckFailImpl(kind, file, line, condition_str, format_str);
+  } else {
+    CheckFailImpl(kind, file, line, condition_str,
+                  llvm::formatv(format_str, std::forward<Ts>(values)...).str());
   }
-
-  // Low-precedence binary operator overload used in check.h macros to flush the
-  // output and exit the program. We do this in a binary operator rather than
-  // the destructor to ensure good debug info and backtraces for errors.
-  [[noreturn]] friend auto operator|(Helper /*helper*/, ExitingStream& stream)
-      -> void {
-    stream.Done();
+}
+#else
+// Use a template parameter for the format string in optimized builds. This will
+// result in essentially every check generating a separate function, but makes
+// the impact on the caller slightly smaller.
+template <TemplateString Kind, TemplateString File, int Line,
+          TemplateString ConditionStr, TemplateString FormatStr, typename... Ts>
+[[noreturn, gnu::cold, clang::noinline, clang::preserve_most]] auto CheckFail(
+    Ts&&... values) -> void {
+  if constexpr (llvm::StringRef(FormatStr).empty() == 0) {
+    CheckFailImpl(Kind.c_str(), File.c_str(), Line, ConditionStr.c_str(),
+                  FormatStr.c_str());
+  } else {
+    CheckFailImpl(
+        Kind.c_str(), File.c_str(), Line, ConditionStr.c_str(),
+        llvm::formatv(FormatStr.c_str(), std::forward<Ts>(values)...).str());
   }
-
- private:
-  [[noreturn]] auto Done() -> void;
-
-  // Whether a separator should be printed if << is used again.
-  bool separator_ = false;
-
-  std::string buffer_str_;
-  llvm::raw_string_ostream buffer_;
-};
+}
+#endif
 
 }  // namespace Carbon::Internal
 
-// Raw exiting stream. This should be used when building forms of exiting
-// macros. It evaluates to a temporary `ExitingStream` object that can be
-// manipulated, streamed into, and then will exit the program.
-#define CARBON_CHECK_INTERNAL_STREAM() \
-  Carbon::Internal::ExitingStream::Helper() | Carbon::Internal::ExitingStream()
+// Dispatch helper between debug and optimized builds for the different
+// formulation of the actual check function.
+#ifndef NDEBUG
+// Pass the format string as a runtime parameter in debug builds.
+#define CARBON_INTERNAL_CHECK_IMPL_CALL(kind, file, line, condition_str, \
+                                        format_str, ...)                 \
+  Carbon::Internal::CheckFail(kind, file, line, condition_str,           \
+                              format_str __VA_OPT__(, ) __VA_ARGS__)
+#else
+// Pass the format string as a template parameter in optimized builds. Note that
+// the template arguments include commas and so require extra parentheses to use
+// within a macro.
```

Comment:
Personally I expect lots of extra parentheses when writing macros, and would just add them in all cases, even the debug case, without any specific comment.